theano/gpuarray/subtensor.py

```diff
@@ -481,6 +481,37 @@ def c_code_cache_version(self):
         return (0,)
 
 
+def check_and_convert_boolean_masks(input, idx_list):
+    """
+    This function checks if the boolean mask arrays in the index have
+    the right shape and converts them to index arrays by calling nonzero.
+    For each boolean mask, we check if the mask has the
+    same shape as the input. This is enforced in NumPy 0.13.0 and
+    newer, but not by earlier versions. If the size is not the same,
+    this method raises an IndexError.
+    """
+    dim_seen = 0
+    out_idx_list = []
+    for index in idx_list:
+        if index is np.newaxis:
+            # skip, does not count as an input dimension
+            out_idx_list.append(index)
+        elif isinstance(index, np.ndarray) and index.dtype == 'bool':
+            for i in xrange(index.ndim):
```

Comment:
I find it confusing as well, but I think what numpy does is this:
- for multi-dimensional integer arrays, it uses the values as indices for a single dimension (and it reshapes the output in that dimension to have the same shape as the index array);
- for boolean arrays, it uses them as a mask, so that if you pass a boolean matrix as the first indexing element, this will be applied to the first two dimensions of the indexed array.

Example of how I think it works:
```python
x = np.arange(16).reshape(2, 2, 2, 2)

# integer index array
a = np.zeros((2, 2), int)
x[a, 0, 0, 0]  # returns [[0, 0], [0, 0]]
# the indices in matrix a applies only to the first dimension of x

# bool mask array
b = np.array([[True, False], [False, True]])
x[b, 0, 0]  # returns [0, 12]
# the first dimension of mask b applies to the first dimension of x,
# the second dimension of b to the second dimension of x,
# so only the third and fourth remain to be indexed
```
For integer arrays, the shape of the index array doesn't matter, because they're used as indices for a single dimension. For boolean arrays, the shape must be the same as that in the corresponding dimensions of the input.